keyv/Ceph.cpp

```diff
@@ -176,118 +78,272 @@ class Ceph : public Plugin
         librados::bufferlist bl;
     };
 
-    typedef stde::hash_map<std::string, AsyncRead> ReadMap;
+    typedef std::unordered_map<std::string, AsyncRead> ReadMap;
     typedef std::deque<librados::AioCompletion*> Writes;
 
     mutable ReadMap _reads;
     Writes _writes;
 
-    bool _fetch(const std::string& key, const size_t sizeHint) const
-    {
-        if (_reads.size() >= _maxPendingOps)
-            return true;
+    bool _fetch(const std::string& key, const size_t sizeHint) const;
+
+    bool _flush(const size_t maxPending);
+
+    librados::bufferlist _get(const std::string& key) const;
+};
 
-        AsyncRead& asyncRead = _reads[key];
-        if (asyncRead.op)
-            return true; // fetch for key already pending
+inline Ceph::Ceph(const servus::URI& uri)
+    : _maxPendingOps(0)
+{
+    const auto poolName = uri.getUserinfo();
+    const auto userName = "client." + poolName;
+    u_int64_t flags = 0;
+    int ret = _cluster.init2(userName.c_str(), uri.getHost().c_str(), flags);
+    if (ret < 0)
+        _throw("Cannot initialize rados cluster", ret);
+
+    auto pos = uri.findQuery("config");
+    if (pos != uri.queryEnd())
+    {
+        ret = _cluster.conf_read_file(pos->second.c_str());
+        if (ret < 0)
+            _throw("Cannot read ceph config '" + pos->second + "'", ret);
+    }
 
-        asyncRead.op.reset(librados::Rados::aio_create_completion());
-        uint64_t size = sizeHint;
-        if (size == 0)
+    pos = uri.findQuery("keyring");
+    if (pos != uri.queryEnd())
+    {
+        ret = _cluster.conf_set("keyring", pos->second.c_str());
+        if (ret < 0)
         {
-            time_t time;
-            const int stat = _context.stat(key, &size, &time);
-            if (stat < 0 || size == 0)
-            {
-                std::cerr << "Stat " << key << " failed: " << ::strerror(-stat)
-                          << std::endl;
-                return false;
-            }
+            _throw("Cannot read ceph keyring '" + pos->second + "'", ret);
         }
+    }
+
+    ret = _cluster.connect();
+    if (ret < 0)
+        _throw("Could not connect rados cluster", ret);
+
+    ret = _cluster.ioctx_create(poolName.c_str(), _context);
+    if (ret < 0)
+        _throw("Could not create io context", ret);
+}
 
-        const int read =
-            _context.aio_read(key, asyncRead.op.get(), &asyncRead.bl, size, 0);
-        if (read >= 0)
+inline Ceph::~Ceph()
+{
+    _context.close();
+    _cluster.shutdown();
+}
+
+inline bool Ceph::handles(const servus::URI& uri)
+{
+    return uri.getScheme() == "ceph";
+}
+
+inline std::string Ceph::getDescription()
+{
+    return "ceph://user@cluster?config=path&keyring=path";
```

Comment:
Can the config and keyring have a default value, e.g., ~/.ceph/config and /keyring.foo?